RIOT-OS/RIOT
# pkg/wakaama: add DTLS support

### Contribution description
This PR adds support for DTLS connections when using LwM2M. The LwM2M client implementation has been rewritten to allow both CoAP and CoAP over DTLS to be used at the same time. An own implementation of the Security Object is introduced, as it needs to handle the interaction with the Credential Manager (credman). For now only Pre-Shared Key mode is supported. DTLS is supported during the communication with LwM2M server and during bootstrapping.

The example application has been modified to use DTLS connections by default. Finally, documentation has been improved.

### Testing procedure
The best way to test this is using the `example/wakaama` application. If you use [Eclipse Leshan](https://github.com/eclipse/leshan) as your LwM2M Server implementation, you should see an indication that the connection is done over DTLS:

![leshan](https://user-images.githubusercontent.com/5381296/112497353-d219b900-8d85-11eb-92f4-440671538e94.png)

![leshan_02](https://user-images.githubusercontent.com/5381296/112497204-ae567300-8d85-11eb-91ae-91feec01aecd.png)

The example can be modified to use plain CoAP connection to the server. That should also work as usual.
```diff
diff --git a/examples/wakaama/lwm2m_cli.c b/examples/wakaama/lwm2m_cli.c
index 0cecb7b73a..8d7392746b 100644
--- a/examples/wakaama/lwm2m_cli.c
+++ b/examples/wakaama/lwm2m_cli.c
@@ -63,8 +63,8 @@ void lwm2m_cli_init(void)
     lwm2m_obj_security_args_t args = {
         .server_id = CONFIG_LWM2M_SERVER_SHORT_ID,
         .server_uri = CONFIG_LWM2M_SERVER_URI,
-        .security_mode = LWM2M_SECURITY_MODE_PRE_SHARED_KEY,
-        .cred = &credential,
+        .security_mode = LWM2M_SECURITY_MODE_NONE,
+        .cred = NULL,
         .is_bootstrap = false, /* set to true when using Bootstrap server */
         .client_hold_off_time = 5,
         .bootstrap_account_timeout = 0
```

### Issues/PRs references
Depends on #16179 
Depends on #16203
